src/System.IO.FileSystem/src/System/IO/Win32FileStream.cs

```diff
@@ -468,10 +469,13 @@ protected override void Dispose(bool disposing)
                 _canRead = false;
                 _canWrite = false;
                 _canSeek = false;
-                // Don't set the buffer to null, to avoid a NullReferenceException
-                // when users have a race condition in their code (ie, they call
-                // Close when calling another method on Stream like Read).
-                //_buffer = null;
+
+                if (_buffer != null)
+                {
+                    ArrayPool<byte>.Shared.Return(_buffer, clearArray: true);
```

Comment:
> We are talking about changing this invariant.

If we care about this to the extreme, even the CopyToAsync changes, and pretty much any use of ```ArrayPool<T>.Shared``` anywhere in corefx, are off-limits.  For example, we hand the buffer to arbitrary ReadAsync/WriteAsync methods, and a bad implementation could easily cache away that buffer for later nefarious purposes.  Now, obviously it's a case of degrees: it's much more likely that code would accidentally Dispose of a FileStream while it's in use than it is to code a ReadAsync/WriteAsync method to store away the provided buffer for later use.

So how much do we care about the invariant?

Unless you disagree, I'll leave the CopyToAsync changes and back out the other ones.  Unfortunate we need to be so conservative in protecting developers from themselves.  Note that @sokket already made some similar changes to DeflateStream in System.IO.Compression... sounds like we should back that out, too? :frowning:.  And cut back or close @sokket's open PR for similar changes to System.IO.